### PR TITLE
Fix concat_ws return null when argument is null

### PR DESCRIPTION
**Pull request by HangyuanLiu**
#918

### CODE REVIEW

File: be/src/exprs/string_functions.cpp [hunk 1]
```diff
@@ -516,17 +516,17 @@ StringVal StringFunctions::concat_ws(
         return strs[0];
     }
 
-    if (strs[0].is_null) {
+    if (strs[0].is_null && 0 == sep.len) {
```
Comment (imay) [inline be/src/exprs/string_functions.cpp]: NULL is different with empty string. 

For NULL, this function return NULL; for empty string, return NOT NULL

you can reference https://dev.mysql.com/doc/refman/8.0/en/string-functions.html#function_concat-ws

**Review by imay — CHANGES_REQUESTED** [review]

File: be/src/exprs/string_functions.cpp (hunk 1 above)
Comment (HangyuanLiu) [reply be/src/exprs/string_functions.cpp]: Fix done, Sorry~

File: be/src/exprs/string_functions.cpp [hunk 2]
```diff
@@ -516,17 +516,17 @@ StringVal StringFunctions::concat_ws(
         return strs[0];
     }
 
-    if (strs[0].is_null) {
+    if (strs[0].is_null && sep.is_null) {
```
Comment (imay) [inline be/src/exprs/string_functions.cpp]: If sep.is_null, function has returned in line 511

File: be/src/exprs/string_functions.cpp [hunk 3]
```diff
@@ -519,14 +510,14 @@ StringVal StringFunctions::concat_ws(
     if (strs[0].is_null) {
```
Comment (imay) [inline be/src/exprs/string_functions.cpp]: put this to for loop

File: be/src/exprs/string_functions.cpp [hunk 4]
```diff
@@ -519,14 +510,14 @@ StringVal StringFunctions::concat_ws(
...
         return StringVal::null();
     }
-    int32_t total_size = strs[0].len;
 
+    int32_t total_size = strs[0].len;
```
Comment (imay) [inline be/src/exprs/string_functions.cpp]: int32_t total_size = 0;

for (int i = 0; i < num_children; ++i)

File: be/src/exprs/string_functions.cpp [hunk 5]
```diff
@@ -535,17 +526,47 @@ StringVal StringFunctions::concat_ws(
...
 }
 
+StringVal StringFunctions::concat_ws(
+        FunctionContext* context, const StringVal& sep, 
+        int num_children, const StringVal* strs) {
+    DCHECK_GE(num_children, 1);
+    if (sep.is_null) {
+        return StringVal::null();
+    }
+
```
Comment (imay) [inline be/src/exprs/string_functions.cpp]: int total_size = 0;

File: be/src/exprs/string_functions.cpp [hunk 6]
```diff
@@ -535,17 +526,47 @@ StringVal StringFunctions::concat_ws(
...
+    int32_t total_size = strs[0].is_null ? 0 : strs[0].len;
+    // Loop once to compute the final size and reserve space.
+    for (int32_t i = 1; i < num_children; ++i) {
```
Comment (imay) [inline be/src/exprs/string_functions.cpp]:
```suggestion
    for (int32_t i = 0; i < num_children; ++i) {
```

File: be/src/exprs/string_functions.cpp [hunk 7]
```diff
@@ -535,17 +526,47 @@ StringVal StringFunctions::concat_ws(
...
+        total_size += strs[i].is_null ? 0 : (sep.len + strs[i].len);
```
Comment (imay) [inline be/src/exprs/string_functions.cpp]: if strs[0] is null, total_size would add a more `sep.len`.

same with your data copy logical

File: be/src/exprs/string_functions.cpp [hunk 8]
```diff
@@ -535,17 +526,47 @@ StringVal StringFunctions::concat_ws(
...
+    }
+
+    // TODO pengyubing
+    // StringVal result = StringVal::create_temp_string_val(context, total_size);
+    StringVal result(context, total_size);
+    uint8_t* ptr = result.ptr;
+    bool not_first = false;
+    // Loop again to append the data.
+    for (int32_t i = 0; i < num_children; ++i) {
+        if (!strs[i].is_null) {
```
Comment (imay) [inline be/src/exprs/string_functions.cpp]:
```suggestion
        if (strs[i].is_null) {
            continue;
        }
```

File: be/src/exprs/string_functions.cpp (hunk 7 above)
Comment (imay) [reply be/src/exprs/string_functions.cpp]: computing size should be equal with appending data

**Review by imay — DISMISSED** [review]

**Review by imay — APPROVED** [review]

**Pull request merged**